File: tokenrazor/core/pruner.py
"""核心剪枝引擎。"""

import re
from typing import List, Optional

from .scanner import RedundancyMatch, Scanner
from .splitter import split_cot_answer, strip_cot_markers


def _clean_text(text: str) -> str:
    """清理剪枝后的文本残留。"""
    # 移除开头的逗号、空格、句号
    text = re.sub(r"^[,，\s。、；;]+", "", text)
    # 移除连续空行
    text = re.sub(r"\n{3,}", "\n\n", text)
    # 清理孤立的标点符号
    text = re.sub(r"[,，]{2,}", "，", text)
    return text.strip()


class PruneResult:
    """一次剪枝操作的结果。"""

    def __init__(
        self,
        original: str,
        pruned: str,
        removed_spans: List[RedundancyMatch],
        stats: dict,
    ):
        self.original = original
        self.pruned = pruned
        self.removed_spans = removed_spans
        self.stats = stats

    @property
    def compression_ratio(self) -> float:
        """压缩率：1.0 = 无压缩，0.5 = 剪掉一半。"""
        orig = self.stats.get("original_tokens", 0)
        pruned = self.stats.get("pruned_tokens", 0)
        if orig == 0:
            return 1.0
        return pruned / orig


class Pruner:
    """CoT 剪枝器。负责编排扫描→剪枝→验证流程。"""

    def __init__(self, strategies: Optional[List[str]] = None):
        self.scanner = Scanner()
        self.strategies = strategies or ["filler", "dead_end"]

    def prune(self, text: str, strict: bool = True) -> PruneResult:
        """对 LLM 输出执行剪枝。

        Args:
            text: LLM 原始输出（含 CoT）
            strict: 严格模式，确保 Answer 部分 100% 不变

        Returns:
            PruneResult
        """
        # 1. 分离 CoT 和 Answer
        cot, answer = split_cot_answer(text)

        if not cot:
            # 无法分离 CoT，保底返回
            return PruneResult(
                original=text,
                pruned=text,
                removed_spans=[],
                stats=self._compute_stats(text, text),
            )

        # 2. 去除 CoT 标记标签，保留内容
        cot_clean = strip_cot_markers(cot)

        # 3. 扫描冗余
        matches = self.scanner.scan_all(cot_clean)

        # 4. 按策略过滤
        matches = [m for m in matches if m.reason.split("+")[0] in self.strategies]

        if not matches:
            return PruneResult(
                original=text,
                pruned=text,
                removed_spans=[],
                stats=self._compute_stats(text, text),
            )

        # 5. 执行剪枝（从后往前删，不影响下标）
        pruned_cot = cot_clean
        removed_spans = []
        for m in reversed(matches):
            pruned_cot = pruned_cot[:m.start] + pruned_cot[m.end:]
            removed_spans.append(m)
        removed_spans.reverse()

        # 6. 清理文本残留
        pruned_cot = _clean_text(pruned_cot)

        # 7. 重建输出
        if answer:
            pruned_text = pruned_cot + "\n\n" + answer
        else:
            pruned_text = pruned_cot

        # 8. 严格模式验证：直接检查 answer 字符串是否完整出现在剪后文本中
        if strict and answer:
            if answer.strip() not in pruned_text:
                # Answer 丢失了，回退到原始
                return PruneResult(
                    original=text,
                    pruned=text,
                    removed_spans=[],
                    stats=self._compute_stats(text, text),
                )

        # 9. 统计
        stats = self._compute_stats(text, pruned_text)

        return PruneResult(
            original=text,
            pruned=pruned_text,
            removed_spans=removed_spans,
            stats=stats,
        )

    def _compute_stats(self, original: str, pruned: str) -> dict:
        from tokenrazor.utils.tokenizer import count_tokens
        orig_tokens = count_tokens(original)
        pruned_tokens = count_tokens(pruned)
        return {
            "original_chars": len(original),
            "pruned_chars": len(pruned),
            "original_tokens": orig_tokens,
            "pruned_tokens": pruned_tokens,
            "saved_tokens": orig_tokens - pruned_tokens,
            "saved_percent": round(100 * (1 - pruned_tokens / orig_tokens), 1) if orig_tokens else 0,
        }

File: rust/src/pruner.rs
/// 剪枝引擎 — 编排扫描→剪枝→验证流程

use crate::scanner::{Scanner, RedundancyMatch};
use crate::tokenizer;

/// 剪枝结果
#[derive(Debug, Clone)]
pub struct PruneResult {
    pub original: String,
    pub pruned: String,
    pub removed_spans: Vec<RedundancyMatch>,
    pub original_tokens: u64,
    pub pruned_tokens: u64,
}

impl PruneResult {
    /// 节省的 token 数
    pub fn saved_tokens(&self) -> u64 {
        self.original_tokens.saturating_sub(self.pruned_tokens)
    }

    /// 节省百分比
    pub fn saved_percent(&self) -> f64 {
        if self.original_tokens == 0 {
            return 0.0;
        }
        (1.0 - self.pruned_tokens as f64 / self.original_tokens as f64) * 100.0
    }
}

/// 剪枝器
pub struct Pruner {
    scanner: Scanner,
    strategies: Vec<String>,
}

impl Default for Pruner {
    fn default() -> Self {
        Self::new(&["filler", "dead_end", "parallel_enum"])
    }
}

impl Pruner {
    pub fn new(strategies: &[&str]) -> Self {
        Self {
            scanner: Scanner::new(),
            strategies: strategies.iter().map(|s| s.to_string()).collect(),
        }
    }

    /// 剪枝文本
    pub fn prune(&self, text: &str) -> PruneResult {
        let strategies: Vec<&str> = self.strategies.iter().map(|s| s.as_str()).collect();

        // 先尝试分离 CoT 和 Answer
        let (cot, answer) = self.split_cot_answer(text);

        if cot.is_empty() || self.scanner.scan_all(&cot, &strategies).is_empty() {
            let tok = tokenizer::count_tokens(text);
            return PruneResult {
                original: text.to_string(),
                pruned: text.to_string(),
                removed_spans: vec![],
                original_tokens: tok,
                pruned_tokens: tok,
            };
        }

        // 扫描冗余
        let matches = self.scanner.scan_all(&cot, &strategies);

        if matches.is_empty() {
            let tok = tokenizer::count_tokens(text);
            return PruneResult {
                original: text.to_string(),
                pruned: text.to_string(),
                removed_spans: vec![],
                original_tokens: tok,
                pruned_tokens: tok,
            };
        }

        // 执行剪枝（从后往前删）
        let mut pruned_cot = cot.to_string();
        let mut removed = Vec::new();
        for m in matches.iter().rev() {
            let new_cot = format!("{}{}", &pruned_cot[..m.start], &pruned_cot[m.end..]);
            pruned_cot = new_cot;
            removed.push(m.clone());
        }
        removed.reverse();

        // 清理文本
        pruned_cot = self.clean_text(&pruned_cot);

        // 重建输出
        let pruned_tok = tokenizer::count_tokens(&pruned_cot);
        let pruned = if answer.is_empty() {
            pruned_cot
        } else {
            format!("{}\n\n{}", pruned_cot, answer)
        };

        PruneResult {
            original: text.to_string(),
            pruned,
            removed_spans: removed,
            original_tokens: tokenizer::count_tokens(text),
            pruned_tokens: pruned_tok,
        }
    }

    /// 分离 CoT 和 Answer（简化版）
    fn split_cot_answer(&self, text: &str) -> (String, String) {
        // 尝试匹配 <thinking> 标签
        if let Some(start) = text.find("<thinking>") {
            if let Some(end) = text.find("</thinking>") {
                let cot = text[start..end + 12].to_string();
                let answer = text[end + 12..].trim().to_string();
                return (cot, answer);
            }
        }

        // 尝试匹配答案关键词
        for keyword in &["\n\n最终答案", "\n\n答案", "\n\nAnswer", "\n\nFinal Answer"] {
            if let Some(pos) = text.find(keyword) {
                let cot = text[..pos].to_string();
                let answer = text[pos..].trim().to_string();
                return (cot, answer);
            }
        }

        // 无标记，全当 CoT
        (text.to_string(), String::new())
    }

    /// 清理文本残留
    fn clean_text(&self, text: &str) -> String {
        let text = text.trim_start_matches(|c: char| c == ',' || c == '，' || c == '。' || c == ' ' || c == '\n');
        let text = text.trim();
        // 移除连续空行
        let re = regex::Regex::new(r"\n{3,}").unwrap();
        re.replace_all(text, "\n\n").to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_prune_filler() {
        let p = Pruner::new(&["filler"]);
        let r = p.prune("让我想想。答案：42");
        assert!(r.saved_tokens() > 0 || r.pruned.len() < r.original.len());
    }

    #[test]
    fn test_prune_plain_text() {
        let p = Pruner::new(&["filler"]);
        let r = p.prune("这是正常的回答内容。");
        // 正常内容不应被修剪
        assert_eq!(r.pruned, r.original);
    }

    #[test]
    fn test_prune_with_thinking_tags() {
        let p = Pruner::new(&["filler", "dead_end"]);
        let text = "<thinking>让我想想。方法一不对。答案：42</thinking>\n最终答案：42";
        let r = p.prune(text);
        // 答案应保留
        assert!(r.pruned.contains("最终答案"));
    }

    #[test]
    fn test_prune_empty() {
        let p = Pruner::new(&["filler"]);
        let r = p.prune("");
        assert_eq!(r.pruned, "");
    }
}
